File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofModels/AgentPick.swift
import Foundation

/// Mirror of an `avatar_picks` row. Ports RN `AgentPick` from
/// `types/agent.ts:160-183`. The `archived_*` JSONB columns are decoded
/// loosely; only fields the iOS UI reads in B13 are typed explicitly.
public struct AgentPick: Codable, Identifiable, Sendable, Hashable {
    public let id: String
    public let avatarId: String
    public let gameId: String
    public let sport: AgentSport
    public let matchup: String
    public let gameDate: String
    public let betType: String
    public let pickSelection: String
    public let odds: String?
    public let units: Double
    public let confidence: Int
    public let reasoningText: String
    public let keyFactors: [String]?
    public let result: PickResultStatus
    public let actualResult: String?
    public let gradedAt: String?
    public let createdAt: String
    /// Raw `ai_decision_trace` JSONB (leaned metrics, rationale, alignment).
    /// Loosely typed — schema drifts across generation versions (v2/v3).
    public let aiDecisionTrace: JSONValue?
    /// Raw `ai_audit_payload` JSONB. V3 payloads carry run_id, steering,
    /// model_response_payload, decision_trace, validator overrides.
    public let aiAuditPayload: JSONValue?

    public enum PickResultStatus: String, Codable, Sendable, Hashable {
        case won
        case lost
        case push
        case pending
    }

    enum CodingKeys: String, CodingKey {
        case id
        case avatarId = "avatar_id"
        case gameId = "game_id"
        case sport
        case matchup
        case gameDate = "game_date"
        case betType = "bet_type"
        case pickSelection = "pick_selection"
        case odds
        case units
        case confidence
        case reasoningText = "reasoning_text"
        case keyFactors = "key_factors"
        case result
        case actualResult = "actual_result"
        case gradedAt = "graded_at"
        case createdAt = "created_at"
        case aiDecisionTrace = "ai_decision_trace"
        case aiAuditPayload = "ai_audit_payload"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(String.self, forKey: .id)
        self.avatarId = try c.decode(String.self, forKey: .avatarId)
        self.gameId = try c.decode(String.self, forKey: .gameId)
        self.sport = (try? c.decode(AgentSport.self, forKey: .sport)) ?? .nfl
        self.matchup = (try? c.decode(String.self, forKey: .matchup)) ?? ""
        self.gameDate = (try? c.decode(String.self, forKey: .gameDate)) ?? ""
        self.betType = (try? c.decode(String.self, forKey: .betType)) ?? ""
        self.pickSelection = (try? c.decode(String.self, forKey: .pickSelection)) ?? ""
        self.odds = try? c.decodeIfPresent(String.self, forKey: .odds)
        if let d = try? c.decode(Double.self, forKey: .units) {
            self.units = d
        } else if let s = try? c.decode(String.self, forKey: .units), let d = Double(s) {
            self.units = d
        } else {
            self.units = 1.0
        }
        self.confidence = (try? c.decode(Int.self, forKey: .confidence)) ?? 3
        self.reasoningText = (try? c.decode(String.self, forKey: .reasoningText)) ?? ""
        self.keyFactors = try? c.decodeIfPresent([String].self, forKey: .keyFactors)
        self.result = (try? c.decode(PickResultStatus.self, forKey: .result)) ?? .pending
        self.actualResult = try? c.decodeIfPresent(String.self, forKey: .actualResult)
        self.gradedAt = try? c.decodeIfPresent(String.self, forKey: .gradedAt)
        self.createdAt = try c.decode(String.self, forKey: .createdAt)
        self.aiDecisionTrace = try? c.decodeIfPresent(JSONValue.self, forKey: .aiDecisionTrace)
        self.aiAuditPayload = try? c.decodeIfPresent(JSONValue.self, forKey: .aiAuditPayload)
    }

    public init(
        id: String,
        avatarId: String,
        gameId: String,
        sport: AgentSport,
        matchup: String,
        gameDate: String,
        betType: String,
        pickSelection: String,
        odds: String?,
        units: Double,
        confidence: Int,
        reasoningText: String,
        keyFactors: [String]?,
        result: PickResultStatus,
        actualResult: String?,
        gradedAt: String?,
        createdAt: String,
        aiDecisionTrace: JSONValue? = nil,
        aiAuditPayload: JSONValue? = nil
    ) {
        self.id = id
        self.avatarId = avatarId
        self.gameId = gameId
        self.sport = sport
        self.matchup = matchup
        self.gameDate = gameDate
        self.betType = betType
        self.pickSelection = pickSelection
        self.odds = odds
        self.units = units
        self.confidence = confidence
        self.reasoningText = reasoningText
        self.keyFactors = keyFactors
        self.result = result
        self.actualResult = actualResult
        self.gradedAt = gradedAt
        self.createdAt = createdAt
        self.aiDecisionTrace = aiDecisionTrace
        self.aiAuditPayload = aiAuditPayload
    }
}

/// Returned by the `get_top_agent_picks_feed_v2` RPC — an `avatar_picks` row
/// joined with the agent's identity + cached perf snapshot. Mirrors RN
/// `TopAgentPickFeedV2Row` in `services/agentPicksService.ts:23-33`.
public struct TopAgentPickFeedRow: Codable, Identifiable, Sendable, Hashable {
    public let id: String
    public let avatarId: String
    public let gameId: String
    public let sport: AgentSport
    public let matchup: String
    public let gameDate: String
    public let betType: String
    public let pickSelection: String
    public let odds: String?
    public let units: Double
    public let confidence: Int
    public let reasoningText: String
    public let result: AgentPick.PickResultStatus
    public let createdAt: String

    public let agentName: String
    public let agentAvatarEmoji: String
    public let agentAvatarColor: String
    public let agentWins: Int
    public let agentLosses: Int
    public let agentPushes: Int
    public let agentNetUnits: Double
    public let agentRank: Int?
    /// Snapshot of the agent's personality at pick time (RPC
    /// `archived_personality`). Drives the strategy chips on the feed card.
    public let archivedPersonality: AgentPersonalityParams?
    /// Agent's current W/L streak (RPC `agent_current_streak`). 0 until the feed
    /// RPC is extended to return it; the form chart degrades to a blank streak.
    public let agentCurrentStreak: Int

    enum CodingKeys: String, CodingKey {
        case id
        case avatarId = "avatar_id"
        case gameId = "game_id"
        case sport
        case matchup
        case gameDate = "game_date"
        case betType = "bet_type"
        case pickSelection = "pick_selection"
        case odds
        case units
        case confidence
        case reasoningText = "reasoning_text"
        case result
        case createdAt = "created_at"
        case agentName = "agent_name"
        case agentAvatarEmoji = "agent_avatar_emoji"
        case agentAvatarColor = "agent_avatar_color"
        case agentWins = "agent_wins"
        case agentLosses = "agent_losses"
        case agentPushes = "agent_pushes"
        case agentNetUnits = "agent_net_units"
        case agentRank = "agent_rank"
        case archivedPersonality = "archived_personality"
        case agentCurrentStreak = "agent_current_streak"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(String.self, forKey: .id)
        self.avatarId = try c.decode(String.self, forKey: .avatarId)
        self.gameId = try c.decode(String.self, forKey: .gameId)
        self.sport = (try? c.decode(AgentSport.self, forKey: .sport)) ?? .nfl
        self.matchup = (try? c.decode(String.self, forKey: .matchup)) ?? ""
        self.gameDate = (try? c.decode(String.self, forKey: .gameDate)) ?? ""
        self.betType = (try? c.decode(String.self, forKey: .betType)) ?? ""
        self.pickSelection = (try? c.decode(String.self, forKey: .pickSelection)) ?? ""
        self.odds = try? c.decodeIfPresent(String.self, forKey: .odds)
        if let d = try? c.decode(Double.self, forKey: .units) {
            self.units = d
        } else if let s = try? c.decode(String.self, forKey: .units), let d = Double(s) {
            self.units = d
        } else {
            self.units = 1.0
        }
        self.confidence = (try? c.decode(Int.self, forKey: .confidence)) ?? 3
        self.reasoningText = (try? c.decode(String.self, forKey: .reasoningText)) ?? ""
        self.result = (try? c.decode(AgentPick.PickResultStatus.self, forKey: .result)) ?? .pending
        self.createdAt = try c.decode(String.self, forKey: .createdAt)
        self.agentName = (try? c.decode(String.self, forKey: .agentName)) ?? "Agent"
        self.agentAvatarEmoji = (try? c.decode(String.self, forKey: .agentAvatarEmoji)) ?? "\u{1F916}"
        self.agentAvatarColor = (try? c.decode(String.self, forKey: .agentAvatarColor)) ?? "#6366f1"
        self.agentWins = (try? c.decode(Int.self, forKey: .agentWins)) ?? 0
        self.agentLosses = (try? c.decode(Int.self, forKey: .agentLosses)) ?? 0
        self.agentPushes = (try? c.decode(Int.self, forKey: .agentPushes)) ?? 0
        if let d = try? c.decode(Double.self, forKey: .agentNetUnits) {
            self.agentNetUnits = d
        } else if let s = try? c.decode(String.self, forKey: .agentNetUnits), let d = Double(s) {
            self.agentNetUnits = d
        } else {
            self.agentNetUnits = 0
        }
        self.agentRank = try? c.decodeIfPresent(Int.self, forKey: .agentRank)
        self.archivedPersonality = try? c.decodeIfPresent(AgentPersonalityParams.self, forKey: .archivedPersonality)
        self.agentCurrentStreak = (try? c.decode(Int.self, forKey: .agentCurrentStreak)) ?? 0
    }
}

File: wagerproof-android-native/core/models/src/main/kotlin/com/wagerproof/core/models/AgentPick.kt
package com.wagerproof.core.models

import com.wagerproof.core.models.serialization.FallbackEnumSerializer
import com.wagerproof.core.models.serialization.FlexibleDoubleOrOneSerializer
import com.wagerproof.core.models.serialization.FlexibleDoubleOrZeroSerializer
import com.wagerproof.core.models.serialization.LossyListSerializer
import com.wagerproof.core.models.serialization.WagerproofJson
import com.wagerproof.core.models.serialization.decodeLossyList
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Mirror of an `avatar_picks` row (RN `AgentPick`). The `archived_*` JSONB
 * columns are decoded loosely; only fields the UI reads are typed explicitly.
 * Tolerant decode: only id/avatarId/gameId are required.
 */
@Serializable
data class AgentPick(
    val id: String,
    @SerialName("avatar_id") val avatarId: String,
    @SerialName("game_id") val gameId: String,
    val sport: AgentSport = AgentSport.NFL,
    val matchup: String = "",
    @SerialName("game_date") val gameDate: String = "",
    @SerialName("bet_type") val betType: String = "",
    @SerialName("pick_selection") val pickSelection: String = "",
    val odds: String? = null,
    @Serializable(with = FlexibleDoubleOrOneSerializer::class)
    val units: Double = 1.0,
    val confidence: Int = 3,
    @SerialName("reasoning_text") val reasoningText: String = "",
    @SerialName("key_factors") val keyFactors: List<String>? = null,
    val result: PickResultStatus = PickResultStatus.PENDING,
    @SerialName("actual_result") val actualResult: String? = null,
    @SerialName("graded_at") val gradedAt: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    /**
     * Raw `ai_decision_trace` JSONB (leaned metrics, rationale, alignment).
     * Loosely typed — schema drifts across generation versions (v2/v3).
     */
    @SerialName("ai_decision_trace") val aiDecisionTrace: JsonElement? = null,
    /** Raw `ai_audit_payload` JSONB (V3 run_id, steering, decision_trace, overrides). */
    @SerialName("ai_audit_payload") val aiAuditPayload: JsonElement? = null,
) {
    @Serializable(with = PickResultStatusSerializer::class)
    enum class PickResultStatus(val raw: String) {
        WON("won"),
        LOST("lost"),
        PUSH("push"),
        PENDING("pending"),
    }

    companion object {
        /**
         * Decode a pick array from raw PostgREST JSON, skipping rows that fail
         * to parse instead of failing the whole batch — one bad row used to
         * blank all of history + performance charts.
         */
        fun decodeLossyArray(json: Json = WagerproofJson, raw: String): List<AgentPick> =
            json.decodeLossyList(serializer(), raw)

        fun decodeLossyArray(json: Json = WagerproofJson, element: JsonElement?): List<AgentPick> =
            json.decodeLossyList(serializer(), element)
    }
}

object PickResultStatusSerializer : FallbackEnumSerializer<AgentPick.PickResultStatus>(
    serialName = "PickResultStatus",
    rawToValue = AgentPick.PickResultStatus.entries.associateBy { it.raw },
    valueToRaw = { it.raw },
    default = AgentPick.PickResultStatus.PENDING,
)

/** Field-level lossy list for embedding pick arrays inside other payloads. */
object AgentPickLossyListSerializer : LossyListSerializer<AgentPick>(AgentPick.serializer())

/**
 * Row returned by the `get_top_agent_picks_feed_v2` RPC — an `avatar_picks`
 * row joined with the agent's identity + cached perf snapshot (RN
 * `TopAgentPickFeedV2Row`). Required: id/avatarId/gameId/createdAt.
 */
@Serializable
data class TopAgentPickFeedRow(
    val id: String,
    @SerialName("avatar_id") val avatarId: String,
    @SerialName("game_id") val gameId: String,
    val sport: AgentSport = AgentSport.NFL,
    val matchup: String = "",
    @SerialName("game_date") val gameDate: String = "",
    @SerialName("bet_type") val betType: String = "",
    @SerialName("pick_selection") val pickSelection: String = "",
    val odds: String? = null,
    @Serializable(with = FlexibleDoubleOrOneSerializer::class)
    val units: Double = 1.0,
    val confidence: Int = 3,
    @SerialName("reasoning_text") val reasoningText: String = "",
    val result: AgentPick.PickResultStatus = AgentPick.PickResultStatus.PENDING,
    @SerialName("created_at") val createdAt: String,
    @SerialName("agent_name") val agentName: String = "Agent",
    @SerialName("agent_avatar_emoji") val agentAvatarEmoji: String = "🤖",
    @SerialName("agent_avatar_color") val agentAvatarColor: String = "#6366f1",
    @SerialName("agent_wins") val agentWins: Int = 0,
    @SerialName("agent_losses") val agentLosses: Int = 0,
    @SerialName("agent_pushes") val agentPushes: Int = 0,
    @SerialName("agent_net_units")
    @Serializable(with = FlexibleDoubleOrZeroSerializer::class)
    val agentNetUnits: Double = 0.0,
    @SerialName("agent_rank") val agentRank: Int? = null,
    /** Snapshot of the agent's personality at pick time — drives feed-card strategy chips. */
    @SerialName("archived_personality") val archivedPersonality: AgentPersonalityParams? = null,
    /** 0 until the feed RPC is extended to return it; form chart degrades to a blank streak. */
    @SerialName("agent_current_streak") val agentCurrentStreak: Int = 0,
) {
    /**
     * Adapt a feed row into the `AgentPick` shape the ticket views consume.
     * The feed RPC omits `key_factors`/`actual_result`/`graded_at` (not shown
     * on the mini ticket), so those stay null.
     */
    val asAgentPick: AgentPick
        get() = AgentPick(
            id = id, avatarId = avatarId, gameId = gameId, sport = sport,
            matchup = matchup, gameDate = gameDate, betType = betType,
            pickSelection = pickSelection, odds = odds, units = units,
            confidence = confidence, reasoningText = reasoningText,
            keyFactors = null, result = result, actualResult = null,
            gradedAt = null, createdAt = createdAt,
        )
}
